### DISCUSSION TITLE
How to customize Autocomplete's dropdown in v7

### DISCUSSION THREAD

**Discussion by theSdev** [mantinedev/mantine#4885] [Q&A] [answered]
I'm migrating to v7, and I'm scratching my head trying to figure out how to customize `<Autocomplete>`'s dropdown items now that the `itemComponent` prop is gone. Any ideas?

Edit: I should probably clarify that by customization I mean having components with multiple children as dropdown items, as opposed to simple strings.

Edit 2: I have the same problem with `nothingFound`, can't find a replacement for that either.

**Comment by rtivital** [comment, marked as answer]
Autocomplete is an opinionated component built on top of [Combobox](https://mantine.dev/core/combobox/) component. It has a limited set of features to cover only the basic use cases. If you need more advanced features, you can build your own component with [Combobox](https://mantine.dev/core/combobox/). You can find examples of custom autocomplete components on [the examples page](https://mantine.dev/combobox/?e=BasicAutocomplete).

**Reply by theSdev** [reply]
I see. It's a bit unfortunate to have to resort to `Combobox` to be able to customize the UI, as it necessitates reimplementing the filtering part, but of course that's not a big deal. Ideally, though, it'd be nice to have the aforementioned props back somehow.

**Reply by frec-hansen** [reply]
+1 on this. This is the major blocker for us to upgrade as well. We have quite a few places where `itemComponent` was used. And to upgrade to V7, we essentially have to rewrite all these components or build a new one that takes the `itemComponent` as a prop and behave like `Select`. Do you think it's possible to add `itemComponent` back as a prop so essentially we can adopt `ComboBox` incrementally as we need?

**Comment by 7iomka** [comment]
@frec-hansen, @rtivital 
I think it's a huge omission for users not to have a section in the documentation with examples of creating wrapper components. Instead, the documentation shows us examples of how to use the component in a specific location with a specific set of local settings and states and mappings.
If in the case of ordinary inputs or buttons to make a wrapper in the local component library was not such a difficult task, although in the case of inputs it is not always a trivial task, then in the case of combobox on the basis of which on the shoulders of the users have taken an impossible burden to build a combine that would work as a number of components in the previous version of the library - without examples of building such components wrappers will be a lot of negative emotions and feelings of hopelessness.

For example, all our UI input components need to supports at least one style: with floating label.
Only for this purpose, I created for each component based on input a local wrapper, for handle focus/blur events and to pass for my custom function the resolved props to get needed info (emotion styles in v6 and classNames, css attributes and data-attributes in v7).

There is an example for Select component wrapper:
```tsx
import type { SelectProps as NativeSelectProps } from '@mantine/core';
import { useProps, Select as NativeSelect } from '@mantine/core';
import type { FocusEvent, RefObject } from 'react';
import { useCallback, useState } from 'react';
import { useUncontrolled } from '@mantine/hooks';
import type { CustomInputVariant } from '../styles';
import { getCustomInputVariants, getResolvedStyleParts } from '../styles';

interface SelectProps extends NativeSelectProps {
  /* outer ref as an alternative for forwardRef */
  inputRef?: RefObject<HTMLInputElement>;
  inputVariant?: CustomInputVariant;
}

const defaultProps = {
  size: 'md',
  inputVariant: 'containedFloatingLabel',
} as const;

const Select = (props: SelectProps) => {
    const resolvedProps = useProps('Select', defaultProps, props);
    const {
      inputRef,
      label,
      value: passedValue,
      onFocus,
      onBlur,
      defaultValue,
      onChange,
      inputVariant,
      classNames,
      style,
      searchable,
      ...rest
    } = resolvedProps;

    const [focused, setFocused] = useState(false);

    const [value, handleChange] = useUncontrolled({
      value: passedValue,
      defaultValue,
      finalValue: null,
      onChange,
    });

    const handleFocus = useCallback(
      (e: FocusEvent<HTMLInputElement>) => {
        if (typeof onFocus === 'function') {
          onFocus(e);
        }
        setFocused(true);
      },
      [onFocus],
    );

    const handleBlur = useCallback(
      (e: FocusEvent<HTMLInputElement>) => {
        if (typeof onBlur === 'function') {
          onBlur(e);
        }
        setFocused(false);
      },
      [onBlur],
    );

    const inputVariantsConfig = getCustomInputVariants('all', {
      containedFloatingLabel: {
        inputProps: resolvedProps,
        floating:
          (value !== null && String(value).trim().length !== 0) || (searchable ? focused : false),
        hasLabel: !!label,
      },
      containedInlineLabel: {
        inputProps: resolvedProps,
        focused,
      },
    });

    const { classes, resolvedStyle, rootAttrs } = getResolvedStyleParts({
      inputVariantsConfig,
      inputVariant,
      classNames,
      style,
      props,
    });

   
  return (
      <NativeSelect
        ref={inputRef}
        classNames={classes}
        style={resolvedStyle}
        label={label}
        value={value}
        onChange={(v) => handleChange(v)}
        onFocus={handleFocus}
        onBlur={handleBlur}
        searchable={searchable}
        {...rest}
        wrapperProps={{ ...rootAttrs, ...rest.wrapperProps }}
      />
    )
}

export type { SelectProps };
export { Select };

```

If there was a page in the documentation with an example of implementing such components at least on the basis of Combobox - it would be an extremely positive action towards the community that wants to upgrade their projects to the latest version of the library and at the same time not break their heads in week-long attempts to switch to it.

Regarding itemComponent specifically, I myself am still at a loss as to which side to approach in order to firstly think about how the inner kitchen of the wrapper component will look like, and secondly - to think whether itemComponent support is needed as the old api, and if not, how the departure from itemComponent would look like with the new api in the context of using the wrapper component.

I have already asked questions here regarding the proper writing of a wrapper component and I was advised to look at how the Mantine components themselves are implemented.
However, in the case of combobox-based components, all we can be content with is 50 examples of specific use cases from which identifying the common features with support for an identical API + additional functionality looks like a difficult task.

**Reply by rtivital** [reply]
I did not quite get the issue with Combobox components. How are they different from other inputs? They support the same Styles API features and their props are documented in the similar manner. 

Also, I think my communication with you here is not really productive, it's pretty hard to understand large chunks of code without proper context, also in a foreign language for both of us. If you need help with migration and want to share feedback, you are welcome to contact me in Discord DM or in [telegram](https://t.me/rtivital).
